src/test/regress/sql/with.sql

```diff
@@ -1083,3 +1083,35 @@ WITH cte AS (
 
 RESET optimizer;
 DROP TABLE d;
+
+-- Check if sharing is disabled for a SegmentGeneral CTE to avoid deadlock if CTE is
+-- executed with 1-gang and joined with Hashed
+SET optimizer = off;
+--start_ignore
+DROP TABLE IF EXISTS d;
+DROP TABLE IF EXISTS r;
+--end_ignore
+
+CREATE TABLE d (a int, b int) DISTRIBUTED BY (a);
+INSERT INTO d VALUES (1,2),(2,3);
+CREATE TABLE r (a int, b int) DISTRIBUTED REPLICATED;
+INSERT INTO r VALUES (1,2),(3,4);
+
+EXPLAIN (COSTS off)
+WITH cte AS (
+    SELECT count(*) a FROM r
+) SELECT * FROM cte JOIN (SELECT * FROM d JOIN cte USING (a) LIMIT 1) d_join_cte USING (a);
+
+WITH cte AS (
+    SELECT count(*) a FROM r
+) SELECT * FROM cte JOIN (SELECT * FROM d JOIN cte USING (a) LIMIT 1) d_join_cte USING (a);
+
+-- Check that a correct plan is generated in case of joining two SegmentGenerals
```

Comment:
But without a patch for this request, a quite good plan with shared scans is generated!

Reply:
We have a choice of 4 variants:
1) Original behavior: generate good plans for joining 2 SegmentGeneral CTEs, and unexecutable plans for joining SegmentGeneral CTE and Hashed
2) Generate an ok plan for joining SegmentGeneral CTE with Hashed, and ineffective plan for joining 2 SegmentGeneral CTEs including excess motions (because of `limit` executed on dispatcher) by enforcing FLOW_REPLICATED
3) Generate absolutely ineffective plans by enforcing Gather Motion after CTE and executing Shared Scan on dispatcher
4) Avoid sharing SegmentGeneral CTE and let the dispatcher decide the amount of segments executing each scan

We don't know how the CTE would be used at the moment of building & sharing the CTE plan. Thus, we can't stick to any variant that would produce the perfect plan for one case and ineffective plan for others. We can't keep Shared scans for the case "if this CTE would be joined with SegmentGeneral" because we don't know if it would.

Considering that replicated tables are usually small, scanning them again is cheaper than running Motions, so the plan generated with patch is not that bad.